docs/randomized_testing.md

```diff
@@ -168,6 +168,535 @@ roughly the same things from the user (which operations to run, how to generate
+    return nil
+}
+```
+
+As a slight tweak, we could re-use `weavertest.Fake` if we wanted to:
+
+```go
+func (w *Workload) Init(_ context.Context, s *sim.Sim) error {
+    w.replies = map[int]int{}
+    s.RegisterFake(weavertest.Fake[B](&fakeb{}))
+    return nil
+}
+```
+
+**Pros.**
+No magic `Fake` functions. The signature of `Init` is also consistent and
+simple. We may add more methods to `sim.Sim` in the future that workloads can
+call.
+
+**Cons.**
+There are now two simulator types. One for the actual simulator, and one passed
+to `Init`. As with Option 4, the simulator does not know ahead of time which
+components are faked, and the `Init` method can return different fakes on every
+invocation.
+
+### Option 7
+
+This option describes an alternative way to specify generators. It can be used
+by all previous options. With this option, an op argument of type `T` implements
+a `Generator[T]` interface:
+
+```go
+type Generator[T any] interface {
+    Generate(*rand.Rand) T
+}
+```
+
+In the code snippet below, for example, `CallA` receives an argument of type
+`myInt`. `myInt` implements `Generator[myInt]`. `CallA` is called on random
+values of `myInt` generated by the `Generate(*rand.Rand) myInt` method.
+
+```go
+type Workload struct {
+    a weaver.Ref[A]
+    b weaver.Ref[B]
+    replies map[int]int
+}
+
+
+type myInt int
+
+func (myInt) Generate(r *rand.Rand) int {
+    return r.Intn(100)
+}
+
+func (w *Workload) Init(context.Context) error {
+    w.replies = map[int]int{}
+    return nil
+}
+
+func (w *Workload) CallA(ctx context.Context, x myInt) error {
+    if y, err := w.a.Get().A(ctx, int(x)); err == nil {
+        w.replies[x] = y
+    }
+    return nil
+}
+
+func (w *Workload) CallB(ctx context.Context) error {
+    w.b.Get().B(ctx)
+    return nil
+}
+```
+
+We can also provide a family of generally useful generators as well. For
+example, `sim.NonNegativeInt`, `sim.SmallInt`, `sim.ReadableString`, and so on.
+Furthermore, we can support primitive types, like `int` and `string`, by using a
+default generator.
+
+**Pros.**
+
+- If a workload has many ops that re-use a type, this approach is much less
+  onerous than repeatedly defining generator methods.
+- The family of generators that we provide can sometimes reduce the amount of
+  work a user has to do.
+- We can extend this API to include other methods besides `Generate`. For
+  example, if the set of all random values is small, `Generate` could return a
+  list of all values, which the simulator may be able to take advantage of.
+  Existing quickcheck libraries also leverage the idea of [Shrinkable
+  types][shrink] to minimize arbitrary types. We could do the same.
+
+**Cons.**
+
+- If a workload has many ops that don't re-use types much, this approach is more
+  onerous that defining generator methods.
+- The `Generator[T]` interface is clunky. A value of type `T` generates a value
+  of type `T`, which is weird. It means that op arguments are doubling as
```

Comment:
I'm still working on addressing the other comments, but I've been thinking more about how to specify generators. The approach I like the most is a Frankenstein combination of the existing proposals. First, we introduce a handful of interfaces: 

```go
type Generator[T any] interface {
    Generate(r *rand.Rand) T
}

type Shrinker[T any] interface {
    Shrink(T) []T
}

type Formatter[T any] interface {
    Format(T) string
}
```

- `Generator[T]` generates values of type `T`.
- `Shrinker[T]` shrinks values, which is used for minimization.
- `Formatter[T]` pretty prints values, which is used for visualization.

I think these interfaces are as simple as possible, without any subtleties or oddities of some of the other proposals. 

Next, workload methods have plain argument types:

```go
type Workload struct { ... }
func (w *Workload) Foo(ctx context.Context, x int, y string) error { ... }
func (w *Workload) Bar(ctx context.Context, z bool) error { ... }
``` 

There is no casting, no `Get` method, and no confusion about a type doubling as its generator. The user then has to define a separate struct with one method for every op in their workload struct. Methods return the `Generator`s used to generate arguments to the ops.

```go
type WorkloadGenerator struct {}
func (WorkloadGenerator) Foo() (Generator[int], Generator[string]) { ... }
func (WorkloadGenerator) Bar() (Generator[bool]) { ... }
```

To make implementing these methods easier, we provide a family of general purpose `Generator`s. For example, `sim.Int`, `sim.String`, `sim.ReadableString`, and so on. We can also provide basic functions to build `Generator`s. For example, `Pick[T any](xs ...T) Generator[T]` returns a `Generator` that randomly returns the provided elements. Here's how we might implement `WorkloadGenerator`: 

```go
type WorkloadGenerator struct {}
func (WorkloadGenerator) Foo() (Generator[int], Generator[string]) {
    return sim.NegativeInt, sim.Pick("a", "b", "c")
}
func (WorkloadGenerator) Bar() (Generator[bool]) {
    return sim.BiasedFlip(0.75)
}
```

Of course, users are free to implement their own custom instances of `Generator[T]` as well. Moreover, a returned `Generator[T]` may optionally implement the `Shrinker[T]` or `Formatter[T]` interfaces. The general purpose `Generator`s we provide all implement the `Shrinker` and `Formatter` interfaces automatically.

If a user finds themselves re-using the same `Generator` across a bunch of methods, they can use helper functions to avoid some of the repetition. Imagine a key-value store based workload, for example: 

```go
func key() Generator[string] { return sim.Pick("a", "b", "c") }
func value() Generator[string] { return sim.ReadableString }

type StoreGenerator struct {}
func (StoreGenerator) Get() Generator[string] { return key() }
func (StoreGenerator) Set() (Generator[string], Generator[string]) { return key(), value() }
func (StoreGenerator) Swap() (Generator[string], Generator[string]) { return key(), key() }
func (StoreGenerator) Delete() Generator[string] { return key() }
``` 

The main downside of this approach is that it requires by far the most typing out of any approach. The more I thought about this though, the less it bothered me. I started to think of it similar to how Go implements errors and how code is littered with `if err != nil { ... }` code. It's more typing, but there's not really any subtleties or gotchas. I started to feel that this approach is the easiest to understand, but the most annoying to type out.

Finally, a small but related point. We've discussed the idea of generators having state. I started to think this is a bad idea and muddies the otherwise simple abstraction of a generator. I think all state should be in the workload, and the generator is left as a stateless and relatively independent entity.